core/adapters/bridge.go

```diff
@@ -16,79 +16,76 @@ import (
 //
 // If the Perform is resumed with a pending RunResult, the RunResult is marked
 // not pending and the RunResult is returned.
-func (ba *Bridge) Perform(input models.RunResult, store *store.Store) models.RunResult {
+func (ba *Bridge) Perform(input models.RunInput, store *store.Store) models.RunOutput {
 	if input.Status.Finished() {
-		return input
+		return models.RunOutput{
+			Data:         input.Data,
+			Status:       input.Status,
+			ErrorMessage: input.ErrorMessage,
+		}
 	} else if input.Status.PendingBridge() {
-		return resumeBridge(input)
+		return models.NewRunOutputInProgress(input.Data)
 	}
 	return ba.handleNewRun(input, store.Config.BridgeResponseURL())
 }
 
-func resumeBridge(input models.RunResult) models.RunResult {
-	input.Status = models.RunStatusInProgress
-	return input
-}
-
-func (ba *Bridge) handleNewRun(input models.RunResult, bridgeResponseURL *url.URL) models.RunResult {
-	if ba.Params == nil {
-		ba.Params = new(models.JSON)
-	}
+func (ba *Bridge) handleNewRun(input models.RunInput, bridgeResponseURL *url.URL) models.RunOutput {
 	var err error
-	if input.Data, err = input.Data.Merge(*ba.Params); err != nil {
-		return models.RunResultError(baRunResultError("handling data param", err))
+	if input.Data, err = input.Data.Merge(ba.Params); err != nil {
+		return models.NewRunOutputError(baRunResultError("handling data param", err))
 	}
 
 	responseURL := bridgeResponseURL
 	if *responseURL != *zeroURL {
-		responseURL.Path += fmt.Sprintf("/v2/runs/%s", input.CachedJobRunID)
+		responseURL.Path += fmt.Sprintf("/v2/runs/%s", input.JobRunID.String())
 	}
 
 	body, err := ba.postToExternalAdapter(input, responseURL)
 	if err != nil {
-		return models.RunResultError(baRunResultError("post to external adapter", err))
+		return models.NewRunOutputError(baRunResultError("post to external adapter", err))
 	}
 
 	return responseToRunResult(body, input)
 }
 
-func responseToRunResult(body []byte, input models.RunResult) models.RunResult {
-	var output models.RunResult
-
+func responseToRunResult(body []byte, input models.RunInput) models.RunOutput {
 	var brr models.BridgeRunResult
 	err := json.Unmarshal(body, &brr)
 	if err != nil {
-		return models.RunResultError(baRunResultError("unmarshaling JSON", err))
-	} else if brr.HasError() {
-		return brr.RunResult
+		return models.NewRunOutputError(baRunResultError("unmarshaling JSON", err))
 	}
 
-	if brr.RunResult.Data.Exists() && !brr.RunResult.Data.IsObject() {
-		output.CompleteWithResult(brr.RunResult.Data.String())
+	if brr.HasError() {
+		return models.NewRunOutputError(brr.GetError())
 	}
 
-	err = output.Merge(brr.RunResult)
-	if err != nil {
-		output.SetError(err)
+	if brr.ExternalPending {
+		return models.NewRunOutputPendingBridge()
 	}
 
-	return output
+	if brr.Data.IsObject() {
+		return models.NewRunOutputComplete(brr.Data)
+	}
+
+	return models.NewRunOutputCompleteWithResult(brr.Data.String())
```

Comment:
Please ignore if you don't find helpful.
This could be done as a switch statement:
```
switch {
case brr.HasError():
  ..
case brr.ExternalPending:
 ..
case brr.Data.IsObject():
 ..
default:
 ...
}
```